### PR TITLE
[AMDGPU] Add InstCombine rule for ballot.i64 intrinsic in wave32 mode.

### PR DESCRIPTION
**Pull request by vpykhtin**
Substitute with zero-extended to i64 ballot.i32 intrinsic.

### CODE REVIEW

**Comment by vpykhtin** [comment]
> I would expect it to fail to select if the type does not match the wave size. Is there a real use case for it?

I would like to use ballot.i64 in both wave modes because I don't want to select bit width in AddressSanitizer code, it's relatively high in the pass chain and wave mode isn't readily available there.

**Comment by arsenm** [comment]
> I would expect it to fail to select if the type does not match the wave size. Is there a real use case for it?

It's a synthetic operation to begin with and has a pretty clear meaning, just 0 the high bits. It's still valid to read exec/exec_hi in wave32 mode, it's just the high bits are always 0. This mirrors the same behavior (such that we are now using this to implement the read_exec builtins)

**Review by arsenm — COMMENTED** [review]
Is this just about code quality? This probably isn't necessary if you consider the pipeline as a whole. I believe InstCombine is already reducing i64 ballot to i32 based on the subtarget's wave size

**Comment by vpykhtin** [comment]
> Is this just about code quality? This probably isn't necessary if you consider the pipeline as a whole. I believe InstCombine is already reducing i64 ballot to i32 based on the subtarget's wave size

Right, but InstCombine is a bit late. This is needed after https://github.com/llvm/llvm-project/pull/68714, otherwise I need special processing of i64 mask in wave32 mode there. I've implemented the TODO with https://github.com/llvm/llvm-project/pull/71556/commits/51578ea440a20a0785e0a081f30ce333a5db7fbf.

**Comment by vpykhtin** [comment]
Ping.

**Comment by arsenm** [comment]
> Right, but InstCombine is a bit late.

I'm confused, InstCombine can't be too late, it runs many times after all IR producers. There's no practical path between the sanitizer passes and the backend where it won't run

**Comment by vpykhtin** [comment]
> I'm confused, InstCombine can't be too late, it runs many times after all IR producers. There's no practical path between the sanitizer passes and the backend where it won't run

I'm sorry, I thought it's machine IR pass. I don't see wavesize type correction for _ballot_ in _GCNTTIImpl::instCombineIntrinsic_, do you beleive it's better to add it there?

**Comment by arsenm** [comment]
> > I'm confused, InstCombine can't be too late, it runs many times after all IR producers. There's no practical path between the sanitizer passes and the backend where it won't run
> 
> I'm sorry, I thought it's machine IR pass. I don't see wavesize type correction for _ballot_ in _GCNTTIImpl::instCombineIntrinsic_, do you beleive it's better to add it there?

Oh yes, it is missing. I recently added it for amdgcn_mbcnt_hi above. We should do the same for ballot

**Comment by vpykhtin** [comment]
Reimplemented using instcombine rule. I had to move i64/wave32 tests to the existing dedicated test.

Please see the differences for individual commits in this branch to follow the changes made.

**Comment by vpykhtin** [comment]
Removed unrelated changes.

File: llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp [hunk 1]
```diff
@@ -2314,9 +2314,8 @@ void AMDGPUDAGToDAGISel::SelectBRCOND(SDNode *N) {
     SDValue VCMP = Cond->getOperand(0);
     auto CC = cast<CondCodeSDNode>(Cond->getOperand(2))->get();
     auto *CRHS = dyn_cast<ConstantSDNode>(Cond->getOperand(1));
-    if ((CC == ISD::SETEQ || CC == ISD::SETNE) && CRHS && CRHS->isZero() &&
-        // TODO: make condition below an assert after fixing ballot bitwidth.
-        VCMP.getValueType().getSizeInBits() == ST->getWavefrontSize()) {
+    if ((CC == ISD::SETEQ || CC == ISD::SETNE) && CRHS && CRHS->isZero()) {
+      assert(VCMP.getValueType().getSizeInBits() == ST->getWavefrontSize());
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:2318]: You are asserting that instcombine has been run? That seems wrong. What about -O0 compiles?

Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp:2318]: Good point. It looks like I need to perform the transformation in both places - instcombiner and SelectionDAG combiner if  I want to keep ballot folding on -O0 but I'm not sure this is really needed.

Also I need to keep original test running without opt and just add runs with opt.

**Comment by vpykhtin** [comment]
- Undo the assert per Jay's comment for -O0 mode.
- Restored the original test and added runs with opt to it.

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp [hunk 2]
```diff
@@ -961,6 +961,18 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
...
+      // %b64 = call i64 ballot.i64(...)
+      // =>
+      // %b32 = call i32 ballot.i32(...)
+      // %b64 = zext i32 %b32 to i64
+      Function *NewF = Intrinsic::getDeclaration(
+          II.getModule(), Intrinsic::amdgcn_ballot, {IC.Builder.getInt32Ty()});
+      CallInst *NewCall = IC.Builder.CreateCall(NewF, {II.getArgOperand(0)});
+      Value *CastedCall = IC.Builder.CreateZExtOrBitCast(NewCall, II.getType());
+      CastedCall->takeName(&II);
+      return IC.replaceInstUsesWith(II, CastedCall);
```
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:974]: Nit: Can just reuse the same value, e.g.
```
      Value *Call = IC.Builder.CreateCall(NewF, {II.getArgOperand(0)});
      Call = IC.Builder.CreateZExtOrBitCast(NewCall, II.getType());
      Call->takeName(&II);
      return IC.replaceInstUsesWith(II, Call);
```

I also think you should be able to use something like `IC.Builder.CreateIntrinsic` ? There should be a function to create an intrinsic call directly.

Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp]: Thanks, changed to `IC.Builder.CreateIntrinsic`. I decided to use it as a temp for `CreateZExtOrBitCast`.

**Comment by vpykhtin** [comment]
I'm going to rebase it on top of https://github.com/llvm/llvm-project/pull/73779 to show test changes.

**Comment by vpykhtin** [comment]
This is ready to be submitted if there're no objections.

**Comment by vpykhtin** [comment]
Gentle ping.

**Review by Flakebi — APPROVED** [review]

**Review by arsenm — COMMENTED** [review]
Description should be adjusted, this isn't really changing the selection anymore

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp [hunk 3]
```diff
@@ -961,6 +961,19 @@ GCNTTIImpl::instCombineIntrinsic(InstCombiner &IC, IntrinsicInst &II) const {
         return IC.replaceInstUsesWith(II, Constant::getNullValue(II.getType()));
       }
     }
+    if (ST->isWave32() && II.getType()->getIntegerBitWidth() == 64) {
+      // %b64 = call i64 ballot.i64(...)
+      // =>
+      // %b32 = call i32 ballot.i32(...)
+      // %b64 = zext i32 %b32 to i64
+      Value *Call = IC.Builder.CreateZExtOrBitCast(
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:969]: Should be CreateZExt, there's no way this can need a bitcast

**Review by arsenm — APPROVED** [review]

**Comment by dyung** [comment]
@vpykhtin the test you modified builtins-amdgcn-wave32.cl seems to be failing on a few bots. Can you take a look?

https://lab.llvm.org/buildbot/#/builders/188/builds/40541
https://lab.llvm.org/buildbot/#/builders/91/builds/21847
https://lab.llvm.org/buildbot/#/builders/98/builds/31671
https://lab.llvm.org/buildbot/#/builders/139/builds/57289

**Comment by vpykhtin** [comment]
> @vpykhtin the test you modified builtins-amdgcn-wave32.cl seems to be failing on a few bots. Can you take a look?

Sure, I'm going to revert and resubmit with the tests updated.

**Pull request merged**